### PR TITLE
fix inferring the correct fn type

### LINKED ISSUE #217: Broken Typescript types when defining options and done
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.18.0

### Plugin version

4.5.0

### Node.js version

18

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

-

### Description

I can't define a plugin with options and done callback

```ts
import fp from "fastify-plugin";

export default fp(function Health(fastify, opts, done) {
  fastify.get("/health", (req, res) => {
    res.send();
  });
});
```

Results in `TS7006: Parameter 'opts' implicitly has an 'any' type.`

### Steps to Reproduce

See description.

Workaround:

```ts
import fp from "fastify-plugin";
import { FastifyPluginCallback } from "fastify";

const plugin: FastifyPluginCallback = function Health(fastify, opts, done) {
  fastify.get("/health", (req, res) => {
    res.send();
  });

  done();
};

export default fp(plugin);

```

### Expected Behavior

_No response_

### PR DESCRIPTION
**Pull request by Uzlopak**
Fixes #217 

I got help from @webstrand. Props to him. 

It should now work.

According to webstrand 

> Typescript isn't currently capable of partial inference
> if you specify any of the type parameters, the rest get filled in with their defaults
> this is why you see some typescript-specific apis doing stuff like
> foo<Bar>()("blah") because the first function call doesn't perform any inference, but typescript lets the second function call infer from its parameters

Maybe needs a new paragraph regarding typing in the readme.md were we basically write, that they should not use the generics but specify the types on the parameters, when calling fastifyPlugin?

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Uzlopak** [comment]
hmm. i think it is false green. Dont merge

**Review by mcollina — APPROVED** [review]
lgtm

File: types/example-async.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  export interface FastifyExampleAsyncPluginOptions extends FastifyExampleAsyncOptions {
+  }
+  export const fastifyExampleAsync: FastifyExampleAsync
+  export { fastifyExampleAsync as default }
+}
+
+declare function fastifyExampleAsync(...params: Parameters<FastifyExampleAsync>): ReturnType<FastifyExampleAsync>
+
+export default fastifyExampleAsync
```
Comment (Uzlopak) [inline types/example-async.test-d.ts:19]:
```suggestion
export default fastifyExampleAsync

```

**Comment by fox1t** [comment]
Hey. This approach is new to me: instead of using polymorphism, we are going with a generic inference. I like it! I am wondering if there are other points in the Fastify ecosystem where we can use the same approach (the same one we did with the infamous triplet and the export = thing when we discovered it)

**Review by fox1t — APPROVED** [review]

**Comment by IlusionDev** [comment]
It is definitely necessary to update the readme, things have started to break the build in my company because of the types of fastify plugin

**Comment by mcollina** [comment]
How? Could you open a fresh issue?

**Comment by Uzlopak** [comment]
@mcollina  Probably mixed use of async and callback. like we had in #220 
Should we add the infomation, that you should not mix both approaches?

**Comment by mcollina** [comment]
> Should we add the infomation, that you should not mix both approaches?

yes

**Comment by Uzlopak** [comment]
created an issue #221

**Review by lilpoozie2005 — APPROVED** [review]

**Pull request merged**